### PR TITLE
pyqt6-webengine: init

### PR DESCRIPTION
**Pull request by nrdxp**
###### Description of changes

Also provides a build of qutebrowser which uses qt6.

~~I have come across an issue with qutebrowser consistently crashing in certain situations, so I'm gonna leave this as a draft unless/until I can get to the bottom of it.~~ 7623600

> Note: Webengine is a very heavy package to build, reviewers may wish to use my cache which already contains the build from this PR. My cache priority is set lower than the official cache.nixos.org, so once the official cache is popullated with this build, it should ignore my cache automatically:
> url: https://cache.nrdxp.dev
> key: cache.nrdxp.dev:KxAfP+PpOl/8UffuH1+lNr35DEsb/irKl7K7C2igbHQ=

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/python-modules/sip/default.nix [hunk 1]
```diff
@@ -2,53 +2,25 @@
...
   # There aren't tests
   doCheck = false;
 
   pythonImportsCheck = [ "sipbuild" ];
 
-  # FIXME: Why isn't this detected automatically?
-  # Needs to be specified in pyproject.toml, e.g.:
-  # [tool.sip.bindings.MODULE]
-  # tags = [PLATFORM_TAG]
-  platform_tag =
```
Comment (LunNova) [inline pkgs/development/python-modules/sip/default.nix:38]: This was getting referenced elsewhere, see ofborg eval error:

```
error: attribute 'platform_tag' missing

       at /var/lib/ofborg/checkout/repo/38dca4e3aa6bca43ea96d2fcc04e8229/mr-est/ofborg-evaluator-2/pkgs/applications/misc/calibre/default.nix:56:94:

           55|       setup/build.py
           56|     sed -i "s/\[tool.sip.bindings.pictureflow\]/[tool.sip.bindings.pictureflow]\ntags = [\"${python3Packages.sip.platform_tag}\"]/g" \
```

**Comment by nrdxp** [comment]
Thanks to upstream for pointing me to a patch that will fix the github crashes in webengine.
https://codereview.qt-project.org/c/qt/qtwebengine/+/436316

I'll apply and push a fix when the build finishes I can confirm it works.

File: pkgs/applications/networking/browsers/qutebrowser/default.nix [hunk 2]
```diff
@@ -1,41 +1,60 @@
...
+buildPythonApplication {
+  inherit pname version;
+
   # the release tarballs are different from the git checkout!
-  src = fetchurl {
+  src = if isQt6 then
+    fetchFromGitHub {
+      owner = "qutebrowser";
+      repo = "qutebrowser";
+      rev = "5e11e6c7d413cf5c77056ba871a545aae1cfd66a";
```
Comment (rnhmjoj) [inline pkgs/applications/networking/browsers/qutebrowser/default.nix:54]: Can you add a comment that says this commit is the [qt6-v2](https://github.com/qutebrowser/qutebrowser/compare/qt6-v2) branch?

File: pkgs/applications/networking/browsers/qutebrowser/default.nix [hunk 3]
```diff
@@ -1,41 +1,60 @@
...
+  fetchzip {
     url = "https://github.com/mozilla/pdf.js/releases/download/v${version}/pdfjs-${version}-dist.zip";
     hash = "sha256-E7t+0AUndrgi4zfJth0w28RmWLqLyXMUCnueNf/gNi4=";
     stripRoot = false;
   };
 
   backendPackage =
-   if backend == "webengine" then python3Packages.pyqtwebengine else
+   if backend == "webengine" then if isQt6 then python3Packages.pyqt6-webengine else python3Packages.pyqtwebengine else
    if backend == "webkit"    then python3Packages.pyqt5_with_qtwebkit else
```
Comment (rnhmjoj) [inline pkgs/applications/networking/browsers/qutebrowser/default.nix:30]: It's very likely webkit will be dropped for the Qt 6 release, so you can remove this altogether.

File: pkgs/applications/networking/browsers/qutebrowser/default.nix [hunk 4]
```diff
@@ -1,41 +1,60 @@
...
    throw ''
      Unknown qutebrowser backend "${backend}".
      Valid choices are qtwebengine (recommended) or qtwebkit.
    '';
 
-in mkDerivationWith python3Packages.buildPythonApplication rec {
+  buildPythonApplication = if isQt6 then python3Packages.buildPythonApplication else mkDerivationWith python3Packages.buildPythonApplication;
+
   pname = "qutebrowser";
   version = "2.5.2";
```
Comment (rnhmjoj) [inline pkgs/applications/networking/browsers/qutebrowser/default.nix:39]: This should be set to something like `unstable-2022-09-16`.

File: pkgs/applications/networking/browsers/qutebrowser/default.nix (hunk 3 above)
Comment (nrdxp) [reply pkgs/applications/networking/browsers/qutebrowser/default.nix:30]: I agree it should be dropped, but we should probably wait until the qt6 work makes it into a release.

File: pkgs/applications/networking/browsers/qutebrowser/default.nix (hunk 2 above)
Comment (nrdxp) [reply pkgs/applications/networking/browsers/qutebrowser/default.nix:55]: done

File: pkgs/applications/networking/browsers/qutebrowser/default.nix (hunk 4 above)
Comment (nrdxp) [reply pkgs/applications/networking/browsers/qutebrowser/default.nix]: done

**Review by LunNova — APPROVED** [review]

File: pkgs/applications/networking/browsers/qutebrowser/default.nix [hunk 5]
```diff
@@ -94,14 +116,15 @@ in mkDerivationWith python3Packages.buildPythonApplication rec {
 
     # Install icons
     for i in 16 24 32 48 64 128 256 512; do
-        install -Dm644 "icons/qutebrowser-''${i}x''${i}.png" \
+        install -Dm644 "qutebrowser/icons/qutebrowser-''${i}x''${i}.png" \
```
Comment (dotlambda) [inline pkgs/applications/networking/browsers/qutebrowser/default.nix:119]: This breaks the build of qutebrowser without Qt6.

File: pkgs/applications/networking/browsers/qutebrowser/default.nix [hunk 6]
```diff
@@ -94,14 +116,15 @@ in mkDerivationWith python3Packages.buildPythonApplication rec {
...
             "$out/share/icons/hicolor/''${i}x''${i}/apps/qutebrowser.png"
     done
-    install -Dm644 icons/qutebrowser.svg \
+    install -Dm644 ${if isQt6 then "qutebrowser/" else ""}icons/qutebrowser.svg \
         "$out/share/icons/hicolor/scalable/apps/qutebrowser.svg"
 
     # Install scripts
     sed -i "s,/usr/bin/,$out/bin/,g" scripts/open_url_in_instance.sh
+    ${if isQt6 then "rm -rf scripts/{testbrowser,dev}" else ""}
```
Comment (dotlambda) [inline pkgs/applications/networking/browsers/qutebrowser/default.nix:127]:
```suggestion
    ${lib.optionalString isQt6 "rm -rf scripts/{testbrowser,dev}"}
```

**Comment by dotlambda** [comment]
Why was this merged so fast without proper review and without the commit messages being properly formatted?

**Pull request merged**